RavenDB-14406 Pass cancellation token to `ReadForMemoryAsync`

karmeli87 opened the pull request.

karmeli87 committed:
RavenDB-14406 Pass cancellation token to `ReadForMemoryAsync`

ayende commented on src/Raven.Client/Http/RavenCommand.cs:
```diff
@@ -145,8 +149,7 @@ public virtual async Task<ResponseDisposeHandling> ProcessResponse(JsonOperation
                 }
 
                 // We do not cache the stream response.
-                using (var uncompressedStream = await RequestExecutor.ReadAsStreamUncompressedAsync(response).ConfigureAwait(false))
```
Where is the call to `ReadAsStreamUncompressedAsync` now?

karmeli87 replied:
This is draft PR so I temperately removed it, since I noticed that `await response.Content.ReadAsStreamAsync()` return `GZipStream`, so it make me think that maybe we don't need that function any longer (but I need to dig into it).

karmeli87 replied:
Maybe I wrong, but to me it seems that `ReadAsStreamUncompressedAsync` was relevant only if the server return zipped response without the client supporting it. 

Is it even possible now? and if yes, why it was relevant only for `Raw` and not for `Object` response type as well.

But let's tackle this in a different issue I guess.

ayende replied:
No, you are correct that the http handler should be in charge of this

karmeli87 committed:
RavenDB-14406 use StreamReaderWithTimeout for additional commands

karmeli87 committed:
RavenDB-14406 handle the server and add stream writer with timeout

karmeli87 committed:
RavenDB-14406 clean up the code

ayende reviewed (changes requested):
Need to change how we do this. And probably need to cosndier why we even need a stream instead instead of using timed cancellation token

karmeli87 committed:
RavenDB-14406 use read/write with timed cancellation token

karmeli87 committed:
RavenDB-14406 use single class for read and write

karmeli87 committed:
RavenDB-14406 add test

ppekrol commented on src/Raven.Client/Util/StreamWithTimeout.cs:
```diff
@@ -0,0 +1,141 @@
+            set => _stream.Position = value;
+        }
+
+        public new void Dispose()
```
You need to override the Dispose(bool)

karmeli87 replied:
done

ppekrol commented on src/Sparrow/Utils/TimeoutManager.cs:
```diff
@@ -111,6 +112,32 @@ private static TimerTaskHolder GetHolderForDuration(uint duration)
             return value;
         }
 
+        public static async Task<T> WaitForTaskCompletion<T>(this Task<T> task, TimeSpan duration, CancellationTokenSource source = null)
```
internal

ppekrol commented on src/Sparrow/Utils/TimeoutManager.cs:
```diff
@@ -111,6 +112,32 @@ private static TimerTaskHolder GetHolderForDuration(uint duration)
+            return await task.ConfigureAwait(false);
+        }
+
+        public static async Task WaitForTaskCompletion(this Task task, TimeSpan duration, CancellationTokenSource source = null)
```
internal

karmeli87 replied:
removed those 2 methods.

ppekrol commented on src/Sparrow/Utils/TimeoutManager.cs:
```diff
@@ -111,6 +112,32 @@ private static TimerTaskHolder GetHolderForDuration(uint duration)
+            if (t == timeoutTask)
+            {
+                source?.Cancel();
+                throw new TimeoutException();
```
Can we put a message here

ppekrol commented on src/Sparrow/Utils/TimeoutManager.cs:
```diff
@@ -111,6 +112,32 @@ private static TimerTaskHolder GetHolderForDuration(uint duration)
+            if (t == timeoutTask)
+            {
+                source?.Cancel();
+                throw new TimeoutException();
```
Can we put a message here

karmeli87 committed:
RavenDB-14406 address PR comments

ppekrol commented:
@ayende can you do final review?

ayende commented on src/Raven.Client/Util/StreamWithTimeout.cs:
```diff
@@ -0,0 +1,141 @@
+            _canBaseStreamTimeoutOnWrite = _stream.CanTimeout && _stream.WriteTimeout < int.MaxValue;
+
+            if (_canBaseStreamTimeoutOnWrite)
+                _writeTimeout = (int?)writeTimeout?.TotalMilliseconds ?? _stream.WriteTimeout;
```
Shouldn't we set the `WriteTimeout` here? If there is a smaller timeout already defined?

karmeli87 replied:
You right, we need to set the inner stream timeouts as well.

About the smaller value, I think that if we specify a timeout is should take precedence over the existing one,

ayende commented on src/Raven.Client/Util/StreamWithTimeout.cs:
```diff
@@ -0,0 +1,141 @@
+            _canBaseStreamTimeoutOnRead = _stream.CanTimeout && _stream.ReadTimeout < int.MaxValue;
+
+            if (_canBaseStreamTimeoutOnRead)
+                _readTimeout = (int?)readTimeout?.TotalMilliseconds ?? _stream.ReadTimeout;
```
Set `ReadTimeout` ?

ayende commented on src/Raven.Client/Util/StreamWithTimeout.cs:
```diff
@@ -0,0 +1,141 @@
+        private Task WriteAsyncWithTimeout(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            _cts?.Dispose();
+            _cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
```
It is fine (and fairly common) to do both read & write operations at the same time.
But you are using the same `_cts` on both

karmeli87 replied:
Done.

karmeli87 committed:
RavenDB-14406 address PR comments

- set write/read timeout of the inner stream
- use seperate tokens for read and write
